Bottom: 0c1f5b6fbbbb5284cb8e96f08d1ba17d08cb0ac9
Top:    726c60bb5bdc68581a2f4a63004ab2cab4bda8b9
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-03-19 19:50:43 +0800

drop vhost_copy_from()

too slow


---

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index ba262ef..9ad9c1e 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -1998,38 +1998,32 @@ static int get_indirect(struct vhost_virtqueue *vq,
 int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 				__virtio16 *indices, u16 num)
 {
-	int ret;
-	u16 last_avail_idx, total, copied;
-	__virtio16 avail_idx;
+	int ret = 0;
+	u16 last_avail_idx, total;
+	__virtio16 avail_idx, head;
 
 	if (unlikely(vhost_get_avail(vq, avail_idx, &vq->avail->idx))) {
 		vq_err(vq, "Failed to access avail idx at %p\n",
 		       &vq->avail->idx);
 		return -EFAULT;
 	}
+	last_avail_idx = vq->last_avail_idx;
 	vq->avail_idx = vhost16_to_cpu(vq, avail_idx);
 	total = vq->avail_idx - vq->last_avail_idx;
-	ret = total = min(num, total);
+	total = min(total, num);
 
-	last_avail_idx = vq->last_avail_idx & (vq->num - 1);
 	while (total) {
-		int ret2;
-
-		copied = vq->num - last_avail_idx;
-		copied = min(total, copied);
-
-		ret2 = vhost_copy_from_user(vq, indices,
-					    &vq->avail->ring[last_avail_idx],
-					    sizeof(avail_idx) * copied);
+		int ret2 = vhost_get_avail(vq, head,
+			   &vq->avail->ring[last_avail_idx & (vq->num - 1)]);
 		if (unlikely(ret2)) {
 			vq_err(vq, "Failed to get descriptors\n");
 			return -EFAULT;
 		}
-
-		indices += copied;
-		total -= copied;
-		last_avail_idx = 0;
+		indices[ret++] = head;
+		--total;
+		++last_avail_idx;
 	}
+	/* FIXME: update used ring here? together with batch dequing? */
 
 	/* Only get avail ring entries after they have been exposed by guest. */
 	smp_rmb();
